src/drivers/imx/sai.c

```diff
@@ -0,0 +1,312 @@
+	trace_sai("SAI: sai_set_config");
+	uint32_t val_cr2 = 0, val_cr4 = 0, val_cr5 = 0;
+	uint32_t mask_cr2 = 0, mask_cr4 = 0, mask_cr5 = 0;
+	/* TODO: this value will be provided by config */
+	uint32_t sywd = 32;
+
+	switch (config->format & SOF_DAI_FMT_FORMAT_MASK) {
+	case SOF_DAI_FMT_I2S:
+		/*
+		 * Frame low, 1clk before data, one word length for frame sync,
+		 * frame sync starts one serial clock cycle earlier,
+		 * that is, together with the last bit of the previous
+		 * data word.
+		 */
+		val_cr2 |= REG_SAI_CR2_BCP;
+		val_cr4 |= REG_SAI_CR4_FSE | REG_SAI_CR4_FSP;
+		val_cr4 |= REG_SAI_CR4_SYWD(sywd);
+		val_cr4 |= REG_SAI_CR4_MF;
+		val_cr4 |= REG_SAI_CR4_FSE;
+		break;
+	case SOF_DAI_FMT_LEFT_J:
+		/*
+		 * Frame high, one word length for frame sync,
+		 * frame sync asserts with the first bit of the frame.
+		 */
+		val_cr2 |= REG_SAI_CR2_BCP;
+		val_cr4 |= REG_SAI_CR4_SYWD(sywd);
+		val_cr4 |= REG_SAI_CR4_MF;
+		break;
+	case SOF_DAI_FMT_DSP_A:
+		/*
+		 * Frame high, 1clk before data, one bit for frame sync,
+		 * frame sync starts one serial clock cycle earlier,
+		 * that is, together with the last bit of the previous
+		 * data word.
+		 */
+		val_cr2 |= REG_SAI_CR2_BCP;
+		val_cr4 |= REG_SAI_CR4_FSE;
+		val_cr4 |= REG_SAI_CR4_SYWD(0U);
+		val_cr4 |= REG_SAI_CR4_MF;
+		val_cr4 |= REG_SAI_CR4_FSE;
+		break;
+	case SOF_DAI_FMT_DSP_B:
+		/*
+		 * Frame high, one bit for frame sync,
+		 * frame sync asserts with the first bit of the frame.
+		 */
+		val_cr2 |= REG_SAI_CR2_BCP;
+		val_cr4 |= REG_SAI_CR4_SYWD(0U);
+		val_cr4 |= REG_SAI_CR4_MF;
+		break;
+	case SOF_DAI_FMT_PDM:
+		val_cr2 |= REG_SAI_CR2_BCP;
+		val_cr4 &= ~REG_SAI_CR4_MF;
+		val_cr4 |= REG_SAI_CR4_MF;
+		break;
+	case SOF_DAI_FMT_RIGHT_J:
+		val_cr4 |= REG_SAI_CR4_SYWD(sywd);
+		val_cr4 |= REG_SAI_CR4_MF;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* DAI clock inversion */
+	switch (config->format & SOF_DAI_FMT_INV_MASK) {
+	case SOF_DAI_FMT_IB_IF:
+		/* Invert both clocks */
+		val_cr2 ^= REG_SAI_CR2_BCP;
+		val_cr4 ^= REG_SAI_CR4_FSP;
+		break;
+	case SOF_DAI_FMT_IB_NF:
+		/* Invert bit clock */
+		val_cr2 ^= REG_SAI_CR2_BCP;
+		break;
+	case SOF_DAI_FMT_NB_IF:
+		/* Invert frame clock */
+		val_cr4 ^= REG_SAI_CR4_FSP;
+		break;
+	case SOF_DAI_FMT_NB_NF:
+		/* Nothing to do for both normal cases */
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* DAI clock master masks */
+	switch (config->format & SOF_DAI_FMT_MASTER_MASK) {
+	case SOF_DAI_FMT_CBS_CFS:
+		break;
+	case SOF_DAI_FMT_CBM_CFM:
+		val_cr2 |= REG_SAI_CR2_MSEL_MCLK1;
+		val_cr2 |= REG_SAI_CR2_BCD_MSTR;
+		val_cr4 |= REG_SAI_CR4_FSD_MSTR;
+		break;
+	case SOF_DAI_FMT_CBS_CFM:
+		val_cr4 |= REG_SAI_CR4_FSD_MSTR;
+		break;
+	case SOF_DAI_FMT_CBM_CFS:
+		val_cr2 |= REG_SAI_CR2_BCD_MSTR;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* set clock divider as (DIV + 1) * 2 */
+	val_cr2 |= SAI_CLOCK_DIV;
+
+	/* for wolfson codec we just have stereo*/
+	val_cr4 |= REG_SAI_CR4_FRSZ(SAI_STERE_CHANNEL_NB);
```

Comment:
Please remove reference to wolfson. People will be confused by this. Just write something like:
```
/* TODO: set number of slots from config */
val_cr4 |= REG_SAI_CR4_FRSZ(SAI_TDM_SLOTS)

```

And then rename  

``#define SAI_STERE_CHANNEL_NB	2``
to
``#define SAI_TDM_SLOTS 2 ``